util/encrypt/ase_layer.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package encrypt
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"encoding/binary"
+	"io"
+	"sync"
+)
+
+const (
+	// blockSize is the encrypt block size in bytes.
+	blockSize = 1024
+	// counterPerBlock represents a encrypt block has 64 aes blocks
+	counterPerBlock = 64
+)
+
+var blockBufPool = sync.Pool{
+	New: func() interface{} { return make([]byte, blockSize) },
+}
+
+// CtrCipher encrypting data using AES in counter mode
+type CtrCipher struct {
+	nonce uint64
+	block cipher.Block
+}
+
+// NewCtrCipher return a CtrCipher
+func NewCtrCipher(key []byte, nonce uint64) (ctr *CtrCipher, err error) {
+	var block cipher.Block
+	block, err = aes.NewCipher(key)
+	if err != nil {
+		return
+	}
+	ctr = new(CtrCipher)
+	ctr.block = block
+	ctr.nonce = nonce
+	return
+}
+
+// stream returns a cipher.Stream be use to encrypts/decrypts
+func (ctr *CtrCipher) stream(counter uint64) cipher.Stream {
+	counterBuf := make([]byte, aes.BlockSize)
+	binary.BigEndian.PutUint64(counterBuf, ctr.nonce)
+	binary.BigEndian.PutUint64(counterBuf[8:], counter)
+	return cipher.NewCTR(ctr.block, counterBuf)
+}
+
+// Writer implements an io.WriteCloser, it encrypt data using AES before writing to the underlying object.
+type Writer struct {
+	err          error
+	w            io.WriteCloser
+	n            int
+	buf          []byte
+	cipherStream cipher.Stream
+}
+
+// NewWriter returns a new Writer which encrypt data using AES before writing to the underlying object.
+func NewWriter(w io.WriteCloser, ctrCipher *CtrCipher) *Writer {
+	writer := &Writer{w: w}
+	writer.buf = make([]byte, blockSize)
+	writer.cipherStream = ctrCipher.stream(0)
+	return writer
+}
+
+// AvailableSize returns how many bytes are unused in the buffer.
+func (w *Writer) AvailableSize() int { return blockSize - w.n }
+
+// Write implements the io.Writer interface.
+func (w *Writer) Write(p []byte) (n int, err error) {
+	for len(p) > w.AvailableSize() && w.err == nil {
+		copiedNum := copy(w.buf[w.n:], p)
+		w.n += copiedNum
+		err = w.Flush()
+		if err != nil {
+			return
+		}
+		n += copiedNum
+		p = p[copiedNum:]
+	}
+	if w.err != nil {
```

Comment:
Put this check to the first line?

Reply:
IMO, we should not put this check  to first line. This check in this line can check the error causes from `w.Flush()` in L87, and the last time `w.Write()`.

Reply:
But we've already checked the error in line 88?

Reply:
Yes, I will fix it.